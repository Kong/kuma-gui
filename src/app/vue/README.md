---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

