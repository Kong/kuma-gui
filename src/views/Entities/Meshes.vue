<template>
  <div class="all-meshes">
    <FrameSkeleton>
      <DataOverview
        :page-size="pageSize"
        :has-error="hasError"
        :is-loading="isLoading"
        :empty-state="empty_state"
        :table-data="tableData"
        :table-data-is-empty="tableDataIsEmpty"
        :next="next"
        @tableAction="tableAction"
        @loadData="loadData($event)"
      >
        <template v-slot:additionalControls>
          <KButton
            class="add-mesh-button"
            appearance="primary"
            size="small"
            :to="{ path: '/wizard/mesh' }"
            @click.native="onCreateClick"
          >
            <span class="custom-control-icon">
              +
            </span>
            Create Mesh
          </KButton>
        </template>
      </DataOverview>
      <Tabs
        v-if="isEmpty === false"
        :has-error="hasError"
        :is-loading="isLoading"
        :tabs="tabs"
        initial-tab-override="overview"
      >
        <template v-slot:tabHeader>
          <div v-if="entity.basicData">
            <h3> Mesh: {{ entity.basicData.name }}</h3>
          </div>
        </template>
        <template v-slot:overview>
          <LabelList
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
          >
            <div>
              <ul>
                <li
                  v-for="(value, key) in entity.basicData"
                  :key="key"
                >
                  <h4 v-if="key === 'creationTime'">
                    Created
                  </h4>
                  <h4 v-else-if="key === 'modificationTime'">
                    Last Modified
                  </h4>
                  <h4 v-else>
                    {{ key }}
                  </h4>
                  <p v-if="key === 'creationTime' || key === 'modificationTime'">
                    {{ value | readableDate }} <em>({{ value | rawDate }})</em>
                  </p>
                  <p v-else>
                    {{ value }}
                  </p>
                </li>
              </ul>
            </div>
            <div v-if="entity.extendedData && entity.extendedData.length">
              <ul>
                <li
                  v-for="(item, key) in entity.extendedData"
                  :key="key"
                >
                  <h4>{{ item.label }}</h4>
                  <p
                    v-if="item.value"
                    class="label-cols"
                  >
                    <span>
                      {{ item.value.type }}
                    </span>
                    <span>
                      {{ item.value.name }}
                    </span>
                  </p>
                  <KBadge
                    v-else
                    size="small"
                    appearance="danger"
                  >
                    Disabled
                  </KBadge>
                </li>
                <li>
                  <h4>Locality Aware Loadbalancing</h4>
                  <p v-if="entity.localityEnabled">
                    <KBadge
                      size="small"
                      appearance="success"
                    >
                      Enabled
                    </KBadge>
                  </p>
                  <KBadge
                    v-else
                    size="small"
                    appearance="danger"
                  >
                    Disabled
                  </KBadge>
                </li>
              </ul>
            </div>
          </LabelList>
        </template>
        <template v-slot:yaml>
          <YamlView
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
            :content="rawEntity"
          />
        </template>
        <template v-slot:resources>
          <LabelList
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
          >
            <div
              v-for="i in countCols"
              :key="i"
            >
              <ul>
                <li
                  v-for="(item, key) in counts.slice((i - 1) * itemsPerCol, i * itemsPerCol)"
                  :key="key"
                >
                  <h4>{{ item.title }}</h4>
                  <p>{{ item.value | formatValue }}</p>
                </li>
              </ul>
            </div>
          </LabelList>
        </template>
      </Tabs>
    </FrameSkeleton>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import { datadogLogs } from '@datadog/browser-logs'

import Kuma from '@/services/kuma'
import { getTableData } from '@/utils/tableDataUtils'
import { getEmptyInsight, getInitialPolicies } from '@/store/reducers/mesh-insights'
import { datadogLogEvents } from '@/datadogEvents'
import { getSome, humanReadableDate, rawReadableDate, stripTimes } from '@/helpers'
import FrameSkeleton from '@/components/Skeletons/FrameSkeleton'
import DataOverview from '@/components/Skeletons/DataOverview'
import Tabs from '@/components/Utils/Tabs'
import YamlView from '@/components/Skeletons/YamlView'
import LabelList from '@/components/Utils/LabelList'
import { PAGE_SIZE_DEFAULT, POLICY_MAP } from '@/consts'

export default {
  name: 'Meshes',
  metaInfo: {
    title: 'Meshes',
  },
  components: {
    FrameSkeleton,
    DataOverview,
    Tabs,
    YamlView,
    LabelList,
  },
  filters: {
    formatValue(value) {
      return value ? value.toLocaleString('en').toString() : 0
    },
    readableDate(value) {
      return humanReadableDate(value)
    },
    rawDate(value) {
      return rawReadableDate(value)
    },
  },
  data() {
    return {
      isLoading: true,
      isEmpty: false,
      hasError: false,
      entityIsLoading: true,
      entityIsEmpty: false,
      entityHasError: false,
      tableDataIsEmpty: false,
      empty_state: {
        title: 'No Data',
        message: 'There are no Meshes present.',
      },
      tableData: {
        headers: [
          { key: 'actions', hideLabel: true },
          { label: 'Name', key: 'name' },
          { label: 'Type', key: 'type' },
        ],
        data: [],
      },
      tabs: [
        {
          hash: '#overview',
          title: 'Overview',
        },
        {
          hash: '#resources',
          title: 'Resources',
        },
        {
          hash: '#yaml',
          title: 'YAML',
        },
      ],
      entity: {},
      rawEntity: {},
      pageSize: PAGE_SIZE_DEFAULT,
      next: null,
      itemsPerCol: 3,
      meshInsight: getEmptyInsight(),
    }
  },
  computed: {
    ...mapGetters({
      featureFlags: 'config/featureFlags',
    }),
    counts() {
      const {
        policies: allPolicies,
        dataplanes: { total },
      } = this.meshInsight

      const policies = {
        ...getInitialPolicies(),
        ...allPolicies,
      }

      return [
        {
          title: 'Data plane proxies',
          value: total,
        },
        {
          title: POLICY_MAP.CircuitBreaker.title,
          value: policies.CircuitBreaker.total,
        },
        {
          title: POLICY_MAP.FaultInjection.title,
          value: policies.FaultInjection.total,
        },
        {
          title: POLICY_MAP.HealthCheck.title,
          value: policies.HealthCheck.total,
        },
        {
          title: POLICY_MAP.ProxyTemplate.title,
          value: policies.ProxyTemplate.total,
        },
        {
          title: POLICY_MAP.TrafficLog.title,
          value: policies.TrafficLog.total,
        },
        {
          title: POLICY_MAP.TrafficPermission.title,
          value: policies.TrafficPermission.total,
        },
        {
          title: POLICY_MAP.TrafficRoute.title,
          value: policies.TrafficRoute.total,
        },
        {
          title: POLICY_MAP.TrafficTrace.title,
          value: policies.TrafficTrace.total,
        },
        {
          title: POLICY_MAP.RateLimit.title,
          value: policies.RateLimit.total,
        },
        {
          title: POLICY_MAP.Retry.title,
          value: policies.Retry.total,
        },
        {
          title: POLICY_MAP.Timeout.title,
          value: policies.Timeout.total,
        },
        {
          title: POLICY_MAP.MeshGateway.title,
          value: policies.MeshGateway.total,
        },
        {
          title: POLICY_MAP.MeshGatewayRoute.title,
          value: policies.MeshGatewayRoute.total,
        },
      ]
    },
    countCols() {
      return Math.ceil(this.counts.length / this.itemsPerCol)
    },
  },
  watch: {
    $route(to, from) {
      this.init()
    },
  },
  beforeMount() {
    this.init()
  },
  methods: {
    init() {
      this.loadData()
    },

    onCreateClick() {
      datadogLogs.logger.info(datadogLogEvents.CREATE_MESH_CLICKED)
    },
    tableAction(ev) {
      const data = ev

      // load the data into the tabs
      this.getEntity(data)
    },
    async loadData(offset = '0') {
      this.isLoading = true
      this.isEmpty = false

      const mesh = this.$route.params.mesh
      let query

      if (mesh !== 'all') {
        query = this.$route.params.mesh
      }

      try {
        const { data, next } = await getTableData({
          getSingleEntity: Kuma.getMesh.bind(Kuma),
          getAllEntities: Kuma.getAllMeshes.bind(Kuma),
          size: this.pageSize,
          offset,
          query,
        })

        // set pagination
        this.next = next

        // set table data
        if (data.length) {
          this.tableData.data = [...data]
          this.tableDataIsEmpty = false
          this.isEmpty = false

          this.getEntity({ name: data[0].name })
        } else {
          this.tableData.data = []
          this.tableDataIsEmpty = true
          this.isEmpty = true
          this.entityIsEmpty = true
        }
      } catch (error) {
        this.hasError = true
        this.isEmpty = true

        console.error(error)
      } finally {
        this.isLoading = false
      }
    },
    getEntity(entity) {
      this.entityIsLoading = true
      this.entityIsEmpty = false
      this.entityHasError = false

      if (entity && entity !== null) {
        return Kuma.getMesh({ name: entity.name })
          .then((response) => {
            if (response) {
              Kuma.getMeshInsights({ name: entity.name }).then((meshInsightResponse) => {
                this.meshInsight = meshInsightResponse
              })

              // get the counts for this mesh

              // const col1 = getSome(response, ['type', 'name', 'creationTime', 'modificationTime'])
              const col1 = getSome(response, ['type', 'name'])

              const formatted = () => {
                const data = Object.entries(getSome(response, ['mtls', 'logging', 'metrics', 'tracing']))
                const newData = []

                data.forEach((i) => {
                  const label = i[0]
                  const subData = i[1] || null

                  /**
                   * If `enabledBackend` is present, this will match it against
                   * the name found in `backends`, get the `type` from there, and
                   * output the defaultBackend name alongside its matched `type`.
                   *
                   * For Tracing, `defaultBackend` is used instead of `enabledBackend`.
                   */

                  if (subData && subData.enabledBackend) {
                    const enabled = subData.enabledBackend
                    const matched = subData.backends.find((obj) => obj.name === enabled)

                    if (matched) {
                      newData.push({
                        label: label,
                        value: {
                          type: matched.type,
                          name: matched.name,
                        },
                      })
                    }
                  } else if (subData && subData.defaultBackend) {
                    const enabled = subData.defaultBackend
                    const matched = subData.backends.find((obj) => obj.name === enabled)

                    if (matched) {
                      newData.push({
                        label: label,
                        value: {
                          type: matched.type,
                          name: matched.name,
                        },
                      })
                    }
                  } else if (subData && subData.backends) {
                    const backends = subData.backends[0]

                    if (backends) {
                      newData.push({
                        label: label,
                        value: {
                          type: backends.type,
                          name: backends.name,
                        },
                      })
                    }
                  } else {
                    newData.push({
                      label: label,
                      value: null,
                    })
                  }
                })

                return newData
              }

              const isRoutingEnabled = () => {
                const { routing } = response

                return routing && routing.localityAwareLoadBalancing
              }

              this.entity = {
                basicData: col1,
                extendedData: formatted(),
                localityEnabled: isRoutingEnabled(),
              }

              // this.rawEntity = response
              this.rawEntity = stripTimes(response)
            } else {
              this.entity = {}
              this.entityIsEmpty = true
            }
          })
          .catch((error) => {
            this.entityHasError = true
            console.error(error)
          })
          .finally(() => {
            setTimeout(() => {
              this.entityIsLoading = false
            }, process.env.VUE_APP_DATA_TIMEOUT)
          })
      } else {
        setTimeout(() => {
          this.entityIsEmpty = true
          this.entityIsLoading = false
        }, process.env.VUE_APP_DATA_TIMEOUT)
      }
    },
  },
}
</script>

<style lang="scss" scoped>
.add-mesh-button {
  background-color: var(--logo-green) !important;
}
</style>
