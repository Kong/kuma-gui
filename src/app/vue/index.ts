import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
