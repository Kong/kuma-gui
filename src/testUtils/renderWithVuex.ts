import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
