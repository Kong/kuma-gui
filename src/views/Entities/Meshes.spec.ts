import { screen } from '@testing-library/vue'
import userEvent from '@testing-library/user-event'
import renderWithVuex from '@/testUtils/renderWithVuex'
import Meshes from './Meshes.vue'

describe('Meshes.vue', () => {
  it('renders snapshot with Resource tab', async () => {
    const { container } = renderWithVuex(Meshes, {
      mocks: {
        $route: {
          params: {
            mesh: 'all',
          },
        },
      },
      stubs: ['router-link'],
    })

    await screen.findByText(/Mesh: default/)

    await userEvent.click(screen.getByText(/Resources/))

    await screen.findByText(/Rate Limits/)

    expect(container).toMatchSnapshot()
  })
})
