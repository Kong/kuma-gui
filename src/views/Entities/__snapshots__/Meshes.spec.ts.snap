// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meshes.vue renders snapshot with Resource tab 1`] = `
<div>
  <div
    class="all-meshes"
  >
    <div
      class="component-frame"
    >
      <div
        class="data-overview"
        data-testid="data-overview"
      >
        <div
          class="data-table-controls mb-2"
        >
          <router-link-stub
            class="k-button add-mesh-button small primary"
            data-v-38fa8ecf=""
            to="[object Object]"
            type="button"
          >
            <span
              class="custom-control-icon"
            >
              
            +
          
            </span>
            
          Create Mesh
        
          </router-link-stub>
           
          <button
            class="k-button ml-2 refresh-button small primary"
            data-v-38fa8ecf=""
            type="button"
          >
            <div
              class="refresh-icon"
              data-v-38fa8ecf=""
            >
              <svg
                data-v-38fa8ecf=""
                viewBox="0 0 36 36"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  data-v-38fa8ecf=""
                  fill="#fff"
                  fill-rule="nonzero"
                >
                  <path
                    d="M18 5.5a12.465 12.465 0 00-8.118 2.995 1.5 1.5 0 001.847 2.363l.115-.095A9.437 9.437 0 0118 8.5l.272.004a9.487 9.487 0 019.07 7.75l.04.246H25a.5.5 0 00-.416.777l4 6a.5.5 0 00.832 0l4-6 .04-.072A.5.5 0 0033 16.5h-2.601l-.017-.15C29.567 10.2 24.294 5.5 18 5.5zM2.584 18.723l-.04.072A.5.5 0 003 19.5h2.6l.018.15C6.433 25.8 11.706 30.5 18 30.5c3.013 0 5.873-1.076 8.118-2.995a1.5 1.5 0 00-1.847-2.363l-.115.095A9.437 9.437 0 0118 27.5l-.272-.004a9.487 9.487 0 01-9.07-7.75l-.041-.246H11a.5.5 0 00.416-.777l-4-6a.5.5 0 00-.832 0l-4 6z"
                    data-v-38fa8ecf=""
                  />
                </g>
              </svg>
            </div>
             
            <span
              data-v-38fa8ecf=""
            >
              Refresh
            </span>
          </button>
        </div>
         
        <div
          class="data-overview-content"
        >
          <div
            class="data-overview-table"
          >
            <table
              class="k-table micro-table has-hover is-clickable"
              data-v-bfeabd48=""
            >
              <thead
                data-v-bfeabd48=""
              >
                <tr
                  data-v-bfeabd48=""
                >
                  <th
                    class=""
                    data-v-bfeabd48=""
                  >
                    <!---->
                  </th>
                  <th
                    class=""
                    data-v-bfeabd48=""
                  >
                     Name 
                  </th>
                  <th
                    class=""
                    data-v-bfeabd48=""
                  >
                     Type 
                  </th>
                </tr>
              </thead>
              <tbody
                data-v-bfeabd48=""
              >
                <tr
                  data-v-bfeabd48=""
                >
                  <td
                    data-v-bfeabd48=""
                  >
                    <a
                      class="data-table-action-link is-active"
                      data-v-bfeabd48=""
                    >
                      <span
                        class="action-link__active-state"
                        data-v-bfeabd48=""
                      >
                        
              ✓
              
                        <span
                          class="sr-only"
                          data-v-bfeabd48=""
                        >
                          
                Selected
              
                        </span>
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    default
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    Mesh
                  </td>
                </tr>
                <tr
                  data-v-bfeabd48=""
                >
                  <td
                    data-v-bfeabd48=""
                  >
                    <a
                      class="data-table-action-link"
                      data-v-bfeabd48=""
                    >
                      <span
                        class="action-link__normal-state"
                        data-v-bfeabd48=""
                      >
                        
              View
            
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    hello-world
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    Mesh
                  </td>
                </tr>
                <tr
                  data-v-bfeabd48=""
                >
                  <td
                    data-v-bfeabd48=""
                  >
                    <a
                      class="data-table-action-link"
                      data-v-bfeabd48=""
                    >
                      <span
                        class="action-link__normal-state"
                        data-v-bfeabd48=""
                      >
                        
              View
            
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    kong-mania-12
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    Mesh
                  </td>
                </tr>
                <tr
                  data-v-bfeabd48=""
                >
                  <td
                    data-v-bfeabd48=""
                  >
                    <a
                      class="data-table-action-link"
                      data-v-bfeabd48=""
                    >
                      <span
                        class="action-link__normal-state"
                        data-v-bfeabd48=""
                      >
                        
              View
            
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    mesh-01
                  </td>
                  <td
                    data-v-bfeabd48=""
                  >
                    Mesh
                  </td>
                </tr>
              </tbody>
            </table>
             
            <div
              class="pagination"
            >
              <!---->
               
              <button
                class="k-button primary"
                data-v-38fa8ecf=""
                type="button"
              >
                
    Next ›
  
              </button>
            </div>
          </div>
           
          <!---->
           
          <!---->
        </div>
         
        <!---->
         
        <!---->
      </div>
       
      <div
        class="tab-container"
        data-testid="tab-container"
      >
        <header
          class="tab__header"
        >
          <div>
            <h3>
               Mesh: default
            </h3>
          </div>
        </header>
         
        <div
          class="tab__content-container"
        >
          <div
            class="k-tabs"
            data-v-a38b664a=""
          >
            <ul
              aria-label="ktabs"
              data-v-a38b664a=""
              role="tablist"
            >
              <li
                aria-controls="panel-0"
                aria-selected="false"
                class="tab-item"
                data-v-a38b664a=""
                id="overview-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Overview
                </a>
              </li>
              <li
                aria-controls="panel-1"
                aria-selected="true"
                class="tab-item active"
                data-v-a38b664a=""
                id="resources-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Resources
                </a>
              </li>
              <li
                aria-controls="panel-2"
                aria-selected="false"
                class="tab-item"
                data-v-a38b664a=""
                id="yaml-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  YAML
                </a>
              </li>
            </ul>
            <div
              aria-labelledby="overview-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-0"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
            <div
              aria-labelledby="resources-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-1"
              role="tabpanel"
              tabindex="0"
            >
              <div
                data-v-a38b664a=""
              >
                <div
                  class="label-list-content"
                >
                  <div
                    class="kong-card noBorder"
                    data-v-0029352b=""
                  >
                    <!---->
                    <!---->
                    <div
                      class="k-card-body"
                      data-v-0029352b=""
                    >
                      <div
                        class="label-list__col-wrapper multi-col"
                        data-v-0029352b=""
                      >
                        <div
                          data-v-0029352b=""
                        >
                          <ul
                            data-v-0029352b=""
                          >
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Data plane proxies
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                13
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Circuit Breakers
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Fault Injections
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                          </ul>
                        </div>
                        <div
                          data-v-0029352b=""
                        >
                          <ul
                            data-v-0029352b=""
                          >
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Health Checks
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Proxy Templates
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Traffic Logs
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                          </ul>
                        </div>
                        <div
                          data-v-0029352b=""
                        >
                          <ul
                            data-v-0029352b=""
                          >
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Traffic Permissions
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                3
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Traffic Routes
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                5
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Traffic Traces
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                          </ul>
                        </div>
                        <div
                          data-v-0029352b=""
                        >
                          <ul
                            data-v-0029352b=""
                          >
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Rate Limits
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Retries
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Timeouts
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                0
                              </p>
                            </li>
                          </ul>
                        </div>
                        <div
                          data-v-0029352b=""
                        >
                          <ul
                            data-v-0029352b=""
                          >
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Mesh Gateways
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                1
                              </p>
                            </li>
                            <li
                              data-v-0029352b=""
                            >
                              <h4
                                data-v-0029352b=""
                              >
                                Mesh Gateway Routes
                              </h4>
                               
                              <p
                                data-v-0029352b=""
                              >
                                1
                              </p>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <!---->
                 
                <!---->
                 
                <!---->
              </div>
            </div>
            <div
              aria-labelledby="yaml-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-2"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
          </div>
           
          <div>
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
